src/Tools/dotnet-monitor/ParameterCapturing/CaptureParametersOperation.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            _profilerChannel = profilerChannel;
+            _endpointInfo = endpointInfo;
+            _logger = logger;
+            _methods = methods;
+            _duration = duration;
+
+            _requestId = Guid.NewGuid();
+        }
+
+        private async Task<bool> CanEndpointProcessRequestsAsync(CancellationToken token)
+        {
+            DiagnosticsClient client = new(_endpointInfo.Endpoint);
+
+            IDictionary<string, string> env = await client.GetProcessEnvironmentAsync(token);
+
+            if (!env.TryGetValue(InProcessFeaturesIdentifiers.EnvironmentVariables.AvailableInfrastructure.ManagedMessaging, out string isManagedMessagingAvailable) ||
+                !env.TryGetValue(InProcessFeaturesIdentifiers.EnvironmentVariables.AvailableInfrastructure.HostingStartup, out string isHostingStartupAvailable))
+            {
+                return false;
+            }
+
+            return ToolIdentifiers.IsEnvVarValueEnabled(isManagedMessagingAvailable) && ToolIdentifiers.IsEnvVarValueEnabled(isHostingStartupAvailable);
+        }
+
+        public async Task ExecuteAsync(TaskCompletionSource<object> startCompletionSource, CancellationToken token)
+        {
+            // Check if the endpoint is capable of responding to our requests
+            if (!await CanEndpointProcessRequestsAsync(token))
+            {
+                throw new MonitoringException(Strings.ErrorMessage_ParameterCapturingNotAvailable);
+            }
+
+            EventParameterCapturingPipelineSettings settings = new()
+            {
+                Duration = Timeout.InfiniteTimeSpan
+            };
+            settings.OnStartedCapturing += OnStartedCapturing;
+            settings.OnStoppedCapturing += OnStoppedCapturing;
+            settings.OnCapturingFailed += OnCapturingFailed;
+            settings.OnServiceStateUpdate += OnServiceStateUpdate;
+            settings.OnUnknownRequestId += OnUnknownRequestId;
+
+            await using EventParameterCapturingPipeline eventTracePipeline = new(_endpointInfo.Endpoint, settings);
+            Task runPipelineTask = eventTracePipeline.StartAsync(token);
+
+            await _profilerChannel.SendMessage(
+                _endpointInfo,
+                new JsonProfilerMessage(IpcCommand.StartCapturingParameters, new StartCapturingParametersPayload
+                {
+                    RequestId = _requestId,
+                    Duration = _duration,
+                    Methods = _methods
+                }),
+                token);
+
+            try
+            {
+                await _capturingStartedCompletionSource.Task.WaitAsync(token).ConfigureAwait(false);
+                startCompletionSource?.TrySetResult(null);
+
+                await _capturingStoppedCompletionSource.Task.WaitAsync(token).ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
```

Comment:
You need call `startCompletionSource.TrySetException` so that callers who don't wait initially wait on the completion of the ExecuteAsync method (e.g. ones that are waiting for it to start) will fail instead of hanging forever.